### ISSUE TITLE
Could not load file or assembly 'DiagnosticServices.HostingStartup, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

### ISSUE THREAD

**Issue by uveta** [dotnet/aspnetcore#17370]
We are having an issue with an Asp.Net Core 2.2 application hosted in an Azure AppService. It is hosted on .Net Core stack and has additional virtual application path configured (besides "/") to point to its _site\wwwroot_ directory, due to requirements imposed by application gateway. It has Application Insights enabled on project level, via imported NuGet and necessary configuration done in Startup.

Whenever an application is deployed via CI pipeline or restarted due to inactivity, first request targeting additional virtual path will bring it down with the following exceptions logged:

> Could not load file or assembly 'DiagnosticServices.HostingStartup, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)\r\n   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)\r\n   at System.Reflection.Assembly.Load(AssemblyName assemblyRef)\r\n   at Microsoft.AspNetCore.Hosting.WebHostBuilder.BuildCommonServices(AggregateException& hostingStartupErrors)

> Could not load file or assembly 'StartupBootstrapper, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)\r\n   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)\r\n   at System.Reflection.Assembly.Load(AssemblyName assemblyRef)\r\n   at Microsoft.AspNetCore.Hosting.WebHostBuilder.BuildCommonServices(AggregateException& hostingStartupErrors)

If a request is sent to this application using the original "/" path, everything works in order and no exceptions are logged. In other case, these exceptions cause our application to terminate and is automatically restarted by AppService. After the restart it starts behaving properly regardless of path used to access it. This however causes an outage of approximately couple of minutes, and is not very comfortable for first user accessing the service after the restart.

I have found similar issue #14547 that is partially describing the same problem, but it seems to be related to Application Insights site extension, which we are not using. Our application also doesn't exhibit this problem when accessed from default path, but only via additional virtual path.

**Comment by javiercn** [comment]
@uveta thanks for contacting us.

The problem you are describing is likely caused by an issue in the application insights SDK. @davidfowl do you know who/where to route this to?

**Comment by anurse** [comment]
@cijothomas Any thoughts here?

**Comment by SgtSpamCan** [comment]
Same, looking for resolution/reason for this fail.  Occurred with a .Net Core 3.1 app to Azure App Service

**Comment by joaovieirabr** [comment]
Same thing here, when using modules="AspNetCoreModuleV2" on web.config. When I change to "AspNetCoreModule", it works. So, I got to make this in every deploy :(

**Comment by uveta** [comment]
> 
> 
> Same thing here, when using modules="AspNetCoreModuleV2" on web.config. When I change to "AspNetCoreModule", it works. So, I got to make this in every deploy :(

I can only confirm that this solution does not solve our issue. After changing web.config in order to use AspNetCoreModule instead of AspNetCoreModuleV2 the service is behaving the same.

**Comment by shirhatti** [comment]
@uveta While I'm not sure what's happening, I have a couple of observations/questions that may help identify what's happening:

1. Hosting Startup failures shouldn't crash your application. We **specifically** don't fail Startup for HostingStartup

1. Can you confirm you haven't enabled the Application Insights site for your app service resource as you've claimed? Here's what the Application Insights blade in the portal might look like if you only enabled App Insights via the SDK and not the site extension.

![image](https://user-images.githubusercontent.com/4734691/71128727-bf87fb80-21a2-11ea-87bf-8c9094d8eda4.png)

**Comment by cweining** [comment]
@uveta - My team owns the site extension that loads DiagnosticServices.HostingStartup and we have not been able to reproduce the issue you are seeing.  I'm interested in the answer to the questions that shirhatti posed above but I think it would be best to have a phone call so we can dig deeper into this problem. If you want to meet with my team, please send us email to serviceprofilerhelp@microsoft.com.  Thanks!

**Comment by uveta** [comment]
@shirhatti 
This is how app insights blade looks like for problematic service
![image](https://user-images.githubusercontent.com/36596808/71153376-425b9580-2239-11ea-956e-81da4b36fcf4.png)

I believe it means we have it enabled via SDK and not site extension. Regarding your first question, I admit I am not 100% sure that application crashes, as I have no exact way to determine when it is running in application service environment. However, two things have led me to this conclusion: since we have logging enabled via elasticsearch, two errors I have included in my report are logged with 'terminal' level. Second, as this issue happens after every new deployment, application service is inaccessible for longer periods of time (around 5-6 minutes), which led me to believe it is restarted in the meantime. This goes for both staging and production slots, which have auto-swap enabled.

@cweining I would love to get in touch with your team, in case we are not able to solve the issue this way. Let me know if my answers to shirhatti are enough for you to pinpoint the issue or not.

**Comment by davidfowl** [comment]
> I believe it means we have it enabled via SDK and not site extension. 

It means you have the site extension installed (even if you are using the SDK in your application).

> However, two things have led me to this conclusion: since we have logging enabled via elasticsearch, two errors I have included in my report are logged with 'terminal' level.

Which ones? The ones above are caught by the framework and continue execution.

> Second, as this issue happens after every new deployment, application service is inaccessible for longer periods of time (around 5-6 minutes), which led me to believe it is restarted in the meantime. This goes for both staging and production slots, which have auto-swap enabled.

That sounds more convincing and if you can reproduce it then we should absolutely look further. It would help to also look at the diagnose and solve problems blade -> search for application event logs and take a look there to find more clues about the restart:

![image](https://user-images.githubusercontent.com/95136/71154934-8a56ca00-21f1-11ea-81c4-bd809d47522d.png)

**Comment by cweining** [comment]
You could also try turning off the profiler and snapshot debugger. There are switches on the App Insights page that you reference above. I'm curious to see if the error and symptoms go away if you disabled those two features. To be sure they are really disabled, you can rename or delete these three app settings:

APPINSIGHTS_PROFILERFEATURE_VERSION
APPINSIGHTS_SNAPSHOTRFEATURE_VERSION
DiagnosticServices_EXTENSION_VERSION

Turning the switches back on will recreate the settings.

**Comment by davidfowl** [comment]
@uveta are you using NLog? We've found one other example of applications crashing if NLog is being used (see https://github.com/aspnet/AspNetCore/issues/14547#issuecomment-569373496). It seems like there are some loggers that fail (I'm guessing unintentionally if anything critical is logged).

**Comment by SidShetye** [comment]
> > Same thing here, when using modules="AspNetCoreModuleV2" on web.config. When I change to "AspNetCoreModule", it works. So, I got to make this in every deploy :(
> 
> I can only confirm that this solution does not solve our issue. After changing web.config in order to use AspNetCoreModule instead of AspNetCoreModuleV2 the service is behaving the same.

How are you guys building with that change? Getting this error

```

Error		The "TransformWebConfig" task failed unexpectedly.
System.Exception: In process hosting is not supported for AspNetCoreModule. Change the AspNetCoreModule to at least AspNetCoreModuleV2.
   at Microsoft.NET.Sdk.Publish.Tasks.WebConfigTransform.SetAspNetCoreHostingModel(String aspNetCoreHostingModelValue, String aspNetCoreModuleName, XElement aspNetCoreElement)
   at Microsoft.NET.Sdk.Publish.Tasks.WebConfigTransform.TransformAspNetCore(XElement aspNetCoreElement, String appName, Boolean configureForAzure, Boolean useAppHost, String extension, String aspNetCoreModuleName, String aspNetCoreHostingModelValue, String projectFullPath)
   at Microsoft.NET.Sdk.Publish.Tasks.WebConfigTransform.Transform(XDocument webConfig, String appName, Boolean configureForAzure, Boolean useAppHost, String extension, String aspNetCoreModuleName, String aspNetCoreHostingModel, String environmentName, String projectFullPath)
   at Microsoft.NET.Sdk.Publish.Tasks.TransformWebConfig.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

**Comment by uveta** [comment]
@davidfowl Sorry for my oversight, we indeed had site extension on. After we switched it off the troublesome exceptions from AppInsights stopped appearing. As a side-effect, this also helped with post-deploy issues, and I can confirm we do not experience outages on our AppService anymore. To answer your second question, we do not use NLog, however we are using Serilog with ElasticSearch sink.

@SidShetye Yes you are right, we have manually updated web.config on AppService after deploying.

To sum it up, our initial issue was solved by turning off site extension. If you would like to continue investigation by taking steps suggested by @cweining and also trying to reproduce outage mentioned by @davidfowl , I would be glad to help.

**Comment by SidShetye** [comment]
@davidfowl Would appreciate if you know of another code or config workaround (for CI/CD friendliness) that doesn't involve manually modifying post-deploy bits? At least till you resolve this? Thanks!

**Comment by shirhatti** [comment]
It seems like there is workaround when using NLog in the linked thread.

As for the benign error from Application Insights, AI has been updated such that this is no longer an issue.

Please feel free to open a new issue if any problems persists
